integration/stability/soak_parallel_rm.sh

```diff
@@ -0,0 +1,168 @@
+SHIM_NAME="${SHIM_NAME:-cc-shim}"
+PROXY_NAME="${PROXY_NAME:-cc-proxy}"
+
+# let's cap the test. If you want to run until you hit the memory limit
+# then just set this to a very large number
+MAX_CONTAINERS=110
+
+function count_containers() {
+	docker ps -qa | wc -l
+}
+
+function check_all_running() {
+	echo Checking ${how_many} containers have all relevant components
+
+	# check what docker thinks
+	how_many_running=$(count_containers)
+
+	if (( ${how_many_running} != ${how_many} )); then
+		echo "Wrong number of containers running (${how_many_running} != ${how_many}) - stopping"
+		exit -1
+	fi
+
+	# check we have the right number of proxy's
+	how_many_proxys=$(ps --no-header -C ${PROXY_NAME} | wc -l)
+	if (( ${how_many_running} >= 1 )); then
+		# If we have any containers, then we expect to have a single proxy
+		if (( ${how_many_proxys} != 1 )); then
+			echo "Wrong number of proxys running (${how_many_running} containers, ${how_many_proxys} proxys) - stopping"
+			exit -1
+		fi
+	else
+		# No containers running, but if we have run before then the proxy may still be running
+		# It is not clear if the proxy quits if there are no containers left
+		if (( ${how_many_proxys} > 1 )); then
+			echo "Wrong number of proxys running (${how_many_running} containers, ${how_many_proxys} proxys) - stopping"
+			exit -1
+		fi
+	fi
+
+	# check we have the right number of shims
+	how_many_shims=$(ps --no-header -C ${SHIM_NAME} | wc -l)
+	# two shim processes per container...
+	if (( ${how_many_running}*2 != ${how_many_shims} )); then
+		echo "Wrong number of shims running (${how_many_running}*2 != ${how_many_shims} - stopping)"
+		exit -1
+	fi
+
+	# check we have the right number of qemu's
+	how_many_qemus=$(ps --no-header -C ${QEMU_NAME} | wc -l)
+	if (( ${how_many_running} != ${how_many_qemus} )); then
+		echo "Wrong number of qemus running (${how_many_running} != ${how_many_qemus} - stopping)"
+		exit -1
+	fi
+
+	# check we have no runtimes running (they should be transient, we should not 'see them')
+	how_many_runtimes=$(ps --no-header -C ${RUNTIME_NAME} | wc -l)
+	if (( ${how_many_runtimes} )); then
+		echo "Wrong number of runtimes running (${how_many_runtimes} - stopping)"
+		exit -1
+	fi
+}
+
+# reported system 'available' memory
+function get_system_avail() {
+	echo $(free -b | head -2 | tail -1 | awk '{print $7}')
+}
+
+function go() {
+	echo "Running..."
+
+	how_many=0
+
+	while true; do {
+		check_all_running
+
+		echo "Run $RUNTIME: $PAYLOAD: $COMMAND"
+		docker run --runtime=${RUNTIME} -tid ${PAYLOAD} ${COMMAND}
+
+		((how_many++))
+		if (( ${how_many} > ${MAX_CONTAINERS} )); then
+			echo "And we have hit the max ${how_many} containers"
+			return
+		fi
+
+		how_much=$(get_system_avail)
+		if (( ${how_much} < ${MEM_CUTOFF} )); then
+			echo "And we are out of memory on container ${how_many} (${how_much} < ${MEM_CUTOFF})"
+			return
+		fi
+	}
+	done
+}
+
+function kill_all_containers() {
+	present=$(docker ps -qa | wc -l)
```

Comment:
Maybe `count_containers()` could be used here

Reply:
oh, good point!